File: Day3/Part1/main.py
# Answer: 17179

def find_largest_numbers_of_bank(strBank):
    def find_largest(list):
        largest = 0
        for val in list:
            if val > largest:
                largest = val
        return largest

    battery_list = list(map(int, list(strBank)))
    possibilities = []

    for index, battery in enumerate(battery_list):
        each_possibilities = []
        for i, b in enumerate(battery_list):
            each_possibilities.append(int(f"{battery}{b}") if index < i else 0)
        possibilities.append(find_largest(each_possibilities))
    return find_largest(possibilities)


banks = open("Day3/input.txt").read().split("\n")
largest_numbers_of_bank = list(map(find_largest_numbers_of_bank, banks))
print(sum(largest_numbers_of_bank))

File: Day3/Part1/main.js
// Answer = 17179
import fs from "fs"

/**
 * Takes url of input and outputs an array containing all inputs
 * @param url Relative Path/URL of input 
 * @returns 
 */
const getInput = async (url) => {
    const data = fs.readFileSync(url, { encoding: 'utf8', flag: 'r' })
    return data.split("\n")
}

const findLargest = (arr) => {
    let largest = 0
    arr.forEach((val) => {
        if (val > largest) {
            largest = val
        }
    })
    return largest
}

(async () => {
    const batteries = await getInput("Day3/input.txt")
    const largeNumbers = batteries.map((strBattery) => { // strBattery = Each Battery
        // try out all possibilities and find the biggest one..
        const batteryArr = strBattery.split("").map((v) => parseInt(v)) // [5,4,9,1,8,9...]

        const possibilities = batteryArr.map((battery, index) => { // Each numbers
            const possibilities = batteryArr.map((b, i) => {
                if (index < i) return parseInt(`${battery}${b}`) // to follow order
                else return 0 // if it is number itself
            })
            return findLargest(possibilities)
        })
        return findLargest(possibilities)
    })
    console.log(largeNumbers.reduce((prev, a) => prev + a, 0))
})()